### PR TITLE
Improved randomness (part 1)

### PR DESCRIPTION
**Pull request by Leemonn**
* **Addressed Issue(s)**: https://github.com/rathena/rathena/issues/7881

* **Server Mode**:  Both

## Description of Pull Request

Current rnd() implementation generates a number between `0` and `SINT32_MAX`. Then a module is applied to check for chance (for example: `rnd() % 100`) this introduces module bias.

At the same time this implementation is way slower than it should because it uses signed integers and involves calculating the module everytime.

In this PR I'm changing the way these random values are calculated to remove the module bias, make code simpler and also make it way faster.

## Comparison
Here is a comparison using [https://quick-bench.com/q/cL2SnlW8Zj7YuyUz-VQxbppGNck](https://quick-bench.com/q/cL2SnlW8Zj7YuyUz-VQxbppGNck)
image

Tested in Linux with no optimizations:
image

Tested in linux with -O2:
image

Tested in linux with -O3:
image

## Important reminder

```c++
rnd()%n == new_random(n)

rnd()%n+x == (rnd()%n) + x
```
## Why is it called `new_random`? Why didn't you replace all `rnd()`?
I decided to split this into 2 or 3 PRs, otherwise we would have 1 big PR with more than 40 files and 500 lines changed.

The name **new_random** is temporary, it allows us to make the transition easier keeping the old implementation of `rnd()` on files that have not been updated yet.

We can remove `rnd()` and improve the other functions on the last PR.

### CODE REVIEW

**Comment by vstumpf** [comment]
Can we use the same signature as the current `rnd()` and just overload it?
```
uint32 rnd(uint32 low, uint32 high) {
   // actual impl
}
uint32 rnd(uint32 high) {
   rnd(0, high);
}
uint32 rnd(void) {
   rnd (0, INT32_MAX);
}
```

this way we don't have to have `rnd_new()`, we can just change `rnd() % X` to `rnd(X)`

**Comment by Leemonn** [comment]
> Can we use the same signature as the current `rnd()` and just overload it?
> 
> ```
> uint32 rnd(uint32 low, uint32 high) {
>    // actual impl
> }
> uint32 rnd(uint32 high) {
>    rnd(0, high);
> }
> uint32 rnd(void) {
>    rnd (0, INT32_MAX);
> }
> ```
> 
> this way we don't have to have `rnd_new()`, we can just change `rnd() % X` to `rnd(X)`

I wanted a different name to avoid any kind of confusion during the update process. But I will change it to `rnd`, I'm already making smaller pull requrests to make it easier to review so having the same name shouldn't be a problem.

**Comment by Lemongrass3110** [comment]
I dont really like calling it rnd() again. The possibility of using rand() instead of rnd() was already too high.
Quite a few of the references that you changed should also not be replaced in the way you did, but improved as well, if we already touch them.
Just commenting this now, before you have to edit it over and over again.

**Comment by Lemongrass3110** [comment]
I will comment it, so that you see what I mean.

File: src/map/elemental.cpp [hunk 1]
```diff
@@ -622,7 +622,7 @@ static int elemental_ai_sub_timer(s_elemental_data *ed, map_session_data *sd, t_
 			return 1;
 		}
 
-		if( battle_check_range(&ed->bl,target,view_range) && rnd()%100 < 2 ) { // 2% chance to cast attack skill.
+		if( battle_check_range(&ed->bl,target,view_range) && rnd(100) < 2 ) { // 2% chance to cast attack skill.
```
Comment (Lemongrass3110) [inline src/map/elemental.cpp:625]: Should use rnd_chance

File: src/map/itemdb.cpp [hunk 2]
```diff
@@ -2872,7 +2872,7 @@ std::shared_ptr<s_item_group_entry> get_random_itemsubgroup(std::shared_ptr<s_it
 	for (size_t j = 0, max = random->data.size() * 3; j < max; j++) {
 		std::shared_ptr<s_item_group_entry> entry = util::umap_random(random->data);
 
-		if (entry->rate == 0 || rnd() % random->total_rate < entry->rate)	// always return entry for rate 0 ('must' item)
+		if (entry->rate == 0 || rnd(random->total_rate) < entry->rate)	// always return entry for rate 0 ('must' item)
```
Comment (Lemongrass3110) [inline src/map/itemdb.cpp:2875]: Should use rnd_chance

File: src/map/itemdb.cpp [hunk 3]
```diff
@@ -4441,7 +4441,7 @@ void s_random_opt_group::apply( struct item& item ){
 		for( size_t j = 0, max = this->slots[static_cast<uint16>(i)].size() * 3; j < max; j++ ){
 			std::shared_ptr<s_random_opt_group_entry> option = util::vector_random( this->slots[static_cast<uint16>(i)] );
 
-			if( rnd() % 10000 < option->chance ){
+			if( rnd(10000) < option->chance ){
```
Comment (Lemongrass3110) [inline src/map/itemdb.cpp:4444]: Should use rnd_chance

File: src/map/itemdb.cpp [hunk 4]
```diff
@@ -4466,7 +4466,7 @@ void s_random_opt_group::apply( struct item& item ){
 
 			std::shared_ptr<s_random_opt_group_entry> option = util::vector_random( this->random_options );
 
-			if( rnd() % 10000 < option->chance ){
+			if( rnd(10000) < option->chance ){
```
Comment (Lemongrass3110) [inline src/map/itemdb.cpp:4469]: Should use rnd_chance

File: src/map/map.cpp [hunk 5]
```diff
@@ -1736,8 +1736,8 @@ int map_search_freecell(struct block_list *src, int16 m, int16 *x,int16 *y, int1
 	}
 
 	while(tries--) {
-		*x = (rx >= 0)?(rnd()%rx2-rx+bx):(rnd()%(mapdata->xs-2)+1);
-		*y = (ry >= 0)?(rnd()%ry2-ry+by):(rnd()%(mapdata->ys-2)+1);
+		*x = (rx >= 0) ? (rnd(rx2) - rx + bx):(rnd(mapdata->xs - 2) + 1);
+		*y = (ry >= 0) ? (rnd(ry2) - ry + by):(rnd(mapdata->ys - 2) + 1);
```
Comment (Lemongrass3110) [inline src/map/map.cpp:1740]: Should be a new rnd_value [bx - rx, bx + rx] and a new rnd_value [1, mapdata->xs[
Same for y.

File: src/map/map.cpp [hunk 6]
```diff
@@ -1870,7 +1870,7 @@ int map_addflooritem(struct item *item, int amount, int16 m, int16 x, int16 y, i
 
 	if (!map_searchrandfreecell(m,&x,&y,flags&2?1:0))
 		return 0;
-	r = rnd();
+	r = rnd(UINT32_MAX);
```
Comment (Lemongrass3110) [inline src/map/map.cpp:1873]: I am too lazy too figure out the math below.
But it should be changed as well and additionally it should also be moved down.

File: src/map/map.cpp [hunk 7]
```diff
@@ -3089,8 +3089,8 @@ int map_random_dir(struct block_list *bl, int16 *x, int16 *y)
 	if (dist < 1) dist =1;
 
 	do {
-		short j = 1 + 2*(rnd()%4); //Pick a random diagonal direction
-		short segment = 1+(rnd()%dist); //Pick a random interval from the whole vector in that direction
+		short j = 1 + 2*(rnd(4)); //Pick a random diagonal direction
```
Comment (Lemongrass3110) [inline src/map/map.cpp:3092]: Should use DIR_* constant.
Also I would prefer to convert that into a static array/vector and pick a random entry from there.

File: src/map/mercenary.cpp [hunk 8]
```diff
@@ -421,7 +421,7 @@ bool mercenary_dead(s_mercenary_data *md) {
 void mercenary_killbonus(s_mercenary_data *md) {
 	std::vector<sc_type> scs = { SC_MERC_FLEEUP, SC_MERC_ATKUP, SC_MERC_HPUP, SC_MERC_SPUP, SC_MERC_HITUP };
 
-	sc_start(&md->bl,&md->bl, util::vector_random(scs), 100, rnd() % 5, 600000);
+	sc_start(&md->bl,&md->bl, util::vector_random(scs), 100, rnd(5), 600000);
```
Comment (Lemongrass3110) [inline src/map/mercenary.cpp:424]: Should use scs.size()

File: src/map/pet.cpp [hunk 9]
```diff
@@ -710,7 +710,7 @@ int pet_attackskill(struct pet_data *pd, int target_id)
 	if (DIFF_TICK(pd->ud.canact_tick, gettick()) > 0)
 		return 0;
 
-	if (rnd()%100 < (pd->a_skill->rate +pd->pet.intimate*pd->a_skill->bonusrate/1000)) { // Skotlex: Use pet's skill
+	if (rnd(100) < (pd->a_skill->rate +pd->pet.intimate*pd->a_skill->bonusrate/1000)) { // Skotlex: Use pet's skill
```
Comment (Lemongrass3110) [inline src/map/pet.cpp:713]: Should use rnd_chance

File: src/map/pet.cpp [hunk 10]
```diff
@@ -782,8 +782,8 @@ int pet_target_check(struct pet_data *pd,struct block_list *bl,int type)
 			rate = 1;
 	}
 
-	if(rnd()%10000 < rate) {
-		if(pd->target_id == 0 || rnd()%10000 < pet_db_ptr->change_target_rate)
+	if(rnd(10000) < rate) {
```
Comment (Lemongrass3110) [inline src/map/pet.cpp:785]: Should use rnd_chance

File: src/map/pet.cpp [hunk 11]
```diff
@@ -782,8 +782,8 @@ int pet_target_check(struct pet_data *pd,struct block_list *bl,int type)
...
+		if(pd->target_id == 0 || rnd(10000) < pet_db_ptr->change_target_rate)
```
Comment (Lemongrass3110) [inline src/map/pet.cpp:786]: Should use rnd_chance

File: src/map/pet.cpp [hunk 12]
```diff
@@ -942,7 +942,7 @@ static int pet_performance(map_session_data *sd, struct pet_data *pd)
 		val = 1;
 
 	pet_stop_walking(pd,2000<<8);
-	clif_pet_performance(pd, rnd()%val + 1);
+	clif_pet_performance(pd, rnd(val) + 1);
```
Comment (Lemongrass3110) [inline src/map/pet.cpp:945]: Should be a new rnd_value [1,val]

File: src/map/pet.cpp [hunk 13]
```diff
@@ -1315,7 +1315,7 @@ int pet_catch_process2(map_session_data* sd, int target_id)
 	if(battle_config.pet_catch_rate != 100)
 		pet_catch_rate = (pet_catch_rate*battle_config.pet_catch_rate)/100;
 
-	if(rnd()%10000 < pet_catch_rate) {
+	if(rnd(10000) < pet_catch_rate) {
```
Comment (Lemongrass3110) [inline src/map/pet.cpp:1318]: Should use rnd_chance

File: src/map/pet.cpp [hunk 14]
```diff
@@ -1702,7 +1702,7 @@ static int pet_randomwalk(struct pet_data *pd,t_tick tick)
 				c += pd->status.speed;
 		}
 
-		pd->next_walktime = tick+rnd()%1000+MIN_RANDOMWALKTIME+c;
+		pd->next_walktime = tick+rnd(1000)+MIN_RANDOMWALKTIME+c;
```
Comment (Lemongrass3110) [inline src/map/pet.cpp:1705]: Should be a new rnd_value [tick+MIN_RANDOMWALKTIME+c,MIN_RANDOMWALKTIME+c+1000[

File: src/map/quest.cpp [hunk 15]
```diff
@@ -775,7 +775,7 @@ void quest_update_objective(map_session_data *sd, struct mob_data* md)
 		for (const auto &it : qi->dropitem) {
 			if (it->mob_id != 0 && it->mob_id != md->mob_id)
 				continue;
-			if (it->rate < 10000 && rnd()%10000 >= it->rate)
+			if (it->rate < 10000 && rnd(10000) >= it->rate)
```
Comment (Lemongrass3110) [inline src/map/quest.cpp:778]: Should use rnd_chance

File: src/map/script.cpp [hunk 16]
```diff
@@ -5592,7 +5592,7 @@ BUILDIN_FUNC(rand)
 	if( range <= 1 )
 		script_pushint(st, min);
 	else
-		script_pushint(st, rnd()%range + min);
+		script_pushint(st, rnd(range) + min);
```
Comment (Lemongrass3110) [inline src/map/script.cpp:5595]: Should be a new rnd_value [min, min+range[

File: src/map/script.cpp [hunk 17]
```diff
@@ -12545,7 +12545,7 @@ BUILDIN_FUNC(homunculus_mutate)
 	if(script_hasdata(st,2))
 		homun_id = script_getnum(st,2);
 	else
-		homun_id = 6048 + (rnd() % 4);
+		homun_id = 6048 + rnd(4);
```
Comment (Lemongrass3110) [inline src/map/script.cpp:12548]: Should be a new rnd_value [MER_EIRA,MER_ELEANOR]

File: src/map/unit.cpp [hunk 18]
```diff
@@ -2625,7 +2625,7 @@ int unit_calc_pos(struct block_list *bl, int tx, int ty, uint8 dir)
 			int i;
 
 			for( i = 0; i < 12; i++ ) {
-				int k = rnd()%8; // Pick a Random Dir
+				int k = rnd(8); // Pick a Random Dir
```
Comment (Lemongrass3110) [inline src/map/unit.cpp:2628]: Should use DIR_* constant.

**Comment by Lemongrass3110** [comment]
I did not comment all of them, but quite many I guess.

**Comment by Leemonn** [comment]
> I did not comment all of them, but quite many I guess.

Yep those are the same that I saw. I don't want to include that refactor in this PR.
I think it's better to do:
`6048 + (rnd() % 4)` --(PR 1)--> `6048 + rnd(4)` --(PR 2)--> `rnd_value(MER_EIRA,MER_ELEANOR)`
Than:
`6048 + (rnd() % 4)` --(PR 1)--> `rnd_value(MER_EIRA,MER_ELEANOR)`

I find it easier to review and detect errors when doing it step by step, also I don't mind making another PR for those refactors. I prefer smaller PRs.

**Comment by Lemongrass3110** [comment]
The problem is that we dont see what usecases we have, if we do not touch it.
Like you wrote yesterday in Discord that rnd_value can be removed.
Just by reviewing your PR I can tell you that we still need it.

So my approach would be to apply the correct functions to the correct places and then improve what is behind that function.
Know what I mean?

**Comment by Leemonn** [comment]
> The problem is that we dont see what usecases we have, if we do not touch it. Like you wrote yesterday in Discord that rnd_value can be removed. Just by reviewing your PR I can tell you that we still need it.
> 
> So my approach would be to apply the correct functions to the correct places and then improve what is behind that function. Know what I mean?

What I meant was that the current `rnd_value` is not needed since we can use `rnd(min, max)`. But yes I understand what you mean.
If you don't mind all those changes in the PR then I will make them, there are A TON of improvements that can be made, some might have a big impact (like the random walk part on mobs, it does a lot of calculations just to randomize the directions).

Do we agree on the new name being `nrand`?

**Comment by Leemonn** [comment]
> I know even for myself there were times I confused the difference between `rand` and `rnd`. Maybe we could stick to something like `ra_rand` or `athena_rand` or something that distinguishes itself from the default `rand` function.

I will replace all calls to `rnd` using other functions. For example:
`rnd%10` or `rnd(10)` ----> `rnd_value(0, 9)`
`rnd%1000 < rate` ----> `rnd_chance(rate, 1000)`

**Comment by Leemonn** [comment]
New commit pushed. Also addresses:
https://github.com/rathena/rathena/issues/7883
https://github.com/rathena/rathena/issues/7884
https://github.com/rathena/rathena/issues/7885

File: src/login/loginclif.cpp [hunk 19]
```diff
@@ -378,7 +378,7 @@ static int logclif_parse_reqkey(int fd, struct login_session_data *sd){
 	RFIFOSKIP(fd,2);
 	{
 		memset(sd->md5key, '\0', sizeof(sd->md5key));
-		sd->md5keylen = (uint16)(12 + rnd() % 4);
+		sd->md5keylen = rnd_value(12u, 15u);
```
Comment (Leemonn) [inline src/login/loginclif.cpp:381]: Is this intended to be in range [12, 15]? or was it meant to be [12, 16] but they implemented it wrong?

File: src/map/map.cpp [hunk 20]
```diff
@@ -1894,8 +1890,8 @@ int map_addflooritem(struct item *item, int amount, int16 m, int16 x, int16 y, i
 
 	memcpy(&fitem->item,item,sizeof(*item));
 	fitem->item.amount = amount;
-	fitem->subx = (r&3)*3+3;
-	fitem->suby = ((r>>2)&3)*3+3;
+	fitem->subx = rnd_value(1, 4) * 3;
+	fitem->suby = rnd_value(1, 4) * 3;
```
Comment (Leemonn) [inline src/map/map.cpp:1894]: suby and subx are not properly randomized, previously the options were very limited. Now they can freely be 3, 6, 9 or 12 independent of each other's value.

File: src/map/pet.cpp [hunk 21]
```diff
@@ -1672,10 +1672,10 @@ static int pet_randomwalk(struct pet_data *pd,t_tick tick)
 			d = 5;
 
 		for(i = 0; i < retrycount; i++) {
-			int r = rnd(), x, y;
+			int x, y;
 
-			x = pd->bl.x+r%(d*2+1)-d;
-			y = pd->bl.y+r/(d*2+1)%(d*2+1)-d;
+			x = pd->bl.x + rnd_value(-d, d);
+			y = pd->bl.y + rnd_value(-d, d);
```
Comment (Leemonn) [inline src/map/pet.cpp:1678]: x and y are now properly randomized, previously the options were very limited. Now they are independent of each other's value.

File: src/map/pet.cpp [hunk 22]
```diff
@@ -1702,7 +1702,7 @@ static int pet_randomwalk(struct pet_data *pd,t_tick tick)
 				c += pd->status.speed;
 		}
 
-		pd->next_walktime = tick+rnd()%1000+MIN_RANDOMWALKTIME+c;
+		pd->next_walktime = tick + MIN_RANDOMWALKTIME + c + rnd_value(0, 999);
```
Comment (Leemonn) [inline src/map/pet.cpp:1705]: Is this code meant to be `tick + MIN_RANDOMWALKTIME + c + rnd_value(0, 999)`?

**Comment by Leemonn** [comment]
@Lemongrass3110 pinging you just in case you didn't see it

**Review by Lemongrass3110 — APPROVED** [review]

File: src/map/mercenary.cpp [hunk 23]
```diff
@@ -421,7 +421,7 @@ bool mercenary_dead(s_mercenary_data *md) {
 void mercenary_killbonus(s_mercenary_data *md) {
 	std::vector<sc_type> scs = { SC_MERC_FLEEUP, SC_MERC_ATKUP, SC_MERC_HPUP, SC_MERC_SPUP, SC_MERC_HITUP };
 
-	sc_start(&md->bl,&md->bl, util::vector_random(scs), 100, rnd() % 5, 600000);
+	sc_start(&md->bl,&md->bl, util::vector_random(scs), 100, rnd_value(1, 5), 600000);
```
[the remark this answers is not in the document]
Comment (aleos89) [reply src/map/mercenary.cpp:424]: I'm afraid I'm not well versed with the Mercenary system. Do you know if a bug report was created prior?

**Pull request merged**